### PR TITLE
Target limit

### PR DESCRIPTION
**Pull request by ivanfelfo**
Feature:
- target limit validation added

### CODE REVIEW

File: spec/requests/targets/targets_create_spec.rb [hunk 1]
```diff
@@ -33,10 +33,36 @@
...
       end
 
-      it 'will return http forbidden' do
+      it 'will return http bad_request' do
         subject
-        expect(response.status).to eq(403)
+        expect(response.status).to eq(400)
       end
+
+      it 'will return error' do
```
Comment (martinjaimem) [inline spec/requests/targets/targets_create_spec.rb:41]:
```suggestion
      it 'returns an error' do
```

File: spec/support/requests/request_spec_helper.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+module RequestSpecHelper
+  # Parse JSON response to ruby hash
```
Comment (martinjaimem) [inline spec/support/requests/request_spec_helper.rb:2]:
```suggestion
```

File: spec/requests/targets/targets_create_spec.rb [hunk 3]
```diff
@@ -33,10 +33,36 @@
...
+    end
+  end
+
+  context 'when user is not logged in' do
+    it 'returns http unauthorized' do
+      subject
+      expect(response).to have_http_status(401)
+    end
+
+    it 'doesn\t change the target count' do
```
Comment (martinjaimem) [inline spec/requests/targets/targets_create_spec.rb:64]:
```suggestion
    it 'doesn\'t change the target count' do
```

File: spec/requests/targets/targets_create_spec.rb [hunk 4]
```diff
@@ -33,10 +33,36 @@
...
+    end
+
+    context 'when current user tries to exceed the target limit' do
+      before { create_list(:target, 10, user: user) }
+
+      it 'doesn\'t save any new target for the user' do
+        subject
+        expect(user.targets.count).to eq(10)
+        expect(json['error']).to eq('error! target limit reached')
+      end
```
Comment (martinjaimem) [inline spec/requests/targets/targets_create_spec.rb:54]:
```suggestion
      it 'doesn\'t save any new target for the user' do
        subject
        expect(user.targets.count).to eq(10)
      end
```

File: spec/requests/targets/targets_create_spec.rb [hunk 5]
```diff
@@ -33,10 +33,36 @@
...
+      it 'will return http bad_request' do
         subject
-        expect(response.status).to eq(403)
+        expect(response.status).to eq(400)
       end
+
+      it 'will return error' do
+        subject
+        expect(json['errors']['latitude'][0]).to eq('is not a number')
+      end
```
Comment (martinjaimem) [inline spec/requests/targets/targets_create_spec.rb:44]:
```suggestion
      end
      
      it 'doesn\t change the target count' do
        expect { subject }.not_to change(Target, :count)
      end
```

File: app/controllers/v1/targets_controller.rb [hunk 6]
```diff
@@ -1,32 +1,31 @@
...
       target.destroy!
+      head :no_content
     end
 
     private
 
+    def check_limit
+      return unless current_v1_user.targets.count >= Target::MAX_TARGETS_PER_USER
+
+      render json: { error: 'error! target limit reached' }, status: :bad_request
```
Comment (martinjaimem) [inline app/controllers/v1/targets_controller.rb:20]:
```suggestion
      render json: { error: 'Target limit reached' }, status: :bad_request
```

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -3,4 +3,9 @@
 class ApplicationController < ActionController::API
   include DeviseTokenAuth::Concerns::SetUserByToken
   rescue_from ActiveRecord::RecordInvalid, with: :render_record_invalid
+
+  def render_record_invalid(exception)
+    logger.info(exception) # for logging
```
Comment (martinjaimem) [inline app/controllers/application_controller.rb:8]:
```suggestion
    logger.info { exception }
```

File: spec/requests/targets/targets_create_spec.rb (hunk 1 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:41]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 4 above)
Comment (martinjaimem) [reply spec/requests/targets/targets_create_spec.rb:54]: Usually, we should do one `expect` per `it`, and be as descriptive as possible in each `it`

File: spec/support/requests/request_spec_helper.rb (hunk 2 above)
Comment (ivanfelfo) [reply spec/support/requests/request_spec_helper.rb:2]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 3 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:64]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 4 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:54]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 5 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:44]: 👍

File: app/controllers/v1/targets_controller.rb (hunk 6 above)
Comment (ivanfelfo) [reply app/controllers/v1/targets_controller.rb:20]: 👍

File: app/controllers/application_controller.rb (hunk 7 above)
Comment (ivanfelfo) [reply app/controllers/application_controller.rb:8]: 👍

**Review by martinjaimem — APPROVED** [review]

File: spec/requests/targets/targets_create_spec.rb [hunk 8]
```diff
@@ -33,10 +33,39 @@
...
+      it 'will return http bad_request' do
         subject
-        expect(response.status).to eq(403)
+        expect(response.status).to eq(400)
       end
+
+      it 'returns an error' do
+        subject
+        expect(json['errors']['latitude'][0]).to eq('is not a number')
+      end
```
Comment (martinjaimem) [inline spec/requests/targets/targets_create_spec.rb:44]:
```suggestion
      end
      
      s
```

File: app/controllers/v1/targets_controller.rb [hunk 9]
```diff
@@ -1,32 +1,31 @@
 module V1
   class TargetsController < ApplicationController
     before_action :authenticate_v1_user!
+    before_action :check_limit, only: :create
```
Comment (juanmferreira93) [inline app/controllers/v1/targets_controller.rb:4]: what's happens on update? For example, you have 10 targets, and I have 5. I update one of my targets and set as yours.. this validation will allow this, and this is not the expected behaviour. I think you need to catch this case

**Review by juanmferreira93 — COMMENTED** [review]
nice job 🚀 , left one comment

**Review by juanmferreira93 — APPROVED** [review]

File: app/models/target.rb [hunk 10]
```diff
@@ -1,9 +1,21 @@
...
   validates :radius, presence: true, numericality: { greater_than: 0 }
   validates :latitude, presence: true, numericality: true
   validates :longitude, presence: true, numericality: true
+
+  validate :user_target_limit
+
+  private
+
+  def user_target_limit
+    return unless !user.nil? && user.targets.count >= Target::MAX_TARGETS_PER_USER
```
Comment (martinjaimem) [inline app/models/target.rb:17]:
```suggestion
    return unless user.present? && user.targets.count >= Target::MAX_TARGETS_PER_USER
```
```suggestion
    return unless !user.nil? && user.targets.count >= Target::MAX_TARGETS_PER_USER
```

File: app/controllers/v1/targets_controller.rb (hunk 9 above)
Comment (ivanfelfo) [reply app/controllers/v1/targets_controller.rb]: 👍

File: app/controllers/v1/targets_controller.rb [hunk 11]
```diff
@@ -4,11 +4,12 @@ class TargetsController < ApplicationController
 
     def create
       @target = current_v1_user.targets.create!(target_params)
-      render_record_valid
+      render json: @target.errors[:user], status: :bad_request unless @target
```
Comment (martinjaimem) [inline app/controllers/v1/targets_controller.rb:7]: is the `unless @target` necessary?

File: app/models/target.rb (hunk 10 above)
Comment (ivanfelfo) [reply app/models/target.rb:17]: 👍

File: app/controllers/v1/targets_controller.rb (hunk 11 above)
Comment (ivanfelfo) [reply app/controllers/v1/targets_controller.rb:7]: yes

**Pull request merged**